File: libs/moqtail-rs/src/model/parameter/setup_option.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::model::{
  common::pair::KeyValuePair,
  error::ParseError,
  parameter::{authorization_token::AuthorizationToken, constant::SetupOptionType},
};
use bytes::{Bytes, BytesMut};
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SetupOption {
  Path { moqt_path: String },
  AuthorizationToken { token: AuthorizationToken },
  MaxAuthTokenCacheSize { max_size: u64 },
  Authority { authority: String },
  MoqtImplementation { info: String },
}
impl SetupOption {
  pub fn new_path(moqt_path: String) -> Self {
    SetupOption::Path { moqt_path }
  }

  /// Raw-QUIC only: the authority component of the `moqt://` URI.
  /// MUST NOT be sent over WebTransport.
  pub fn new_authority(authority: String) -> Self {
    SetupOption::Authority { authority }
  }

  pub fn new_auth_token(token: AuthorizationToken) -> Self {
    SetupOption::AuthorizationToken { token }
  }

  pub fn new_max_auth_token_cache_size(max_size: u64) -> Self {
    SetupOption::MaxAuthTokenCacheSize { max_size }
  }

  pub fn new_moqt_implementation(info: String) -> Self {
    SetupOption::MoqtImplementation { info }
  }

  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut bytes = BytesMut::new();
    match self {
      Self::Path { moqt_path } => {
        let data = moqt_path.as_bytes();
        let kvp =
          KeyValuePair::try_new_bytes(SetupOptionType::Path as u64, Bytes::copy_from_slice(data))?;
        let slice = kvp.serialize()?;
        bytes.extend_from_slice(&slice);
      }
      Self::AuthorizationToken { token } => match token.serialize() {
        Ok(payload_bytes) => {
          let kvp =
            KeyValuePair::try_new_bytes(SetupOptionType::AuthorizationToken as u64, payload_bytes)?;
          let slice = kvp.serialize()?;
          bytes.extend_from_slice(&slice);
        }
        Err(e) => {
          return Err(ParseError::Other {
            context: "SetupOption::serialize",
            msg: e.to_string(),
          });
        }
      },
      Self::MaxAuthTokenCacheSize { max_size } => {
        let kvp =
          KeyValuePair::try_new_varint(SetupOptionType::MaxAuthTokenCacheSize as u64, *max_size)?;
        let slice = kvp.serialize()?;
        bytes.extend_from_slice(&slice);
      }
      Self::MoqtImplementation { info } => {
        let data = info.as_bytes();
        let kvp = KeyValuePair::try_new_bytes(
          SetupOptionType::MoqtImplementation as u64,
          Bytes::copy_from_slice(data),
        )?;
        let slice = kvp.serialize()?;
        bytes.extend_from_slice(&slice);
      }
      Self::Authority { authority } => {
        let data = authority.as_bytes();
        let kvp = KeyValuePair::try_new_bytes(
          SetupOptionType::Authority as u64,
          Bytes::copy_from_slice(data),
        )?;
        let slice = kvp.serialize()?;
        bytes.extend_from_slice(&slice);
      }
    }
    Ok(bytes.freeze())
  }
  pub fn deserialize(kvp: &KeyValuePair) -> Result<SetupOption, ParseError> {
    match kvp {
      KeyValuePair::VarInt { type_value, value } => {
        let type_value = SetupOptionType::try_from(*type_value)?;
        match type_value {
          SetupOptionType::MaxAuthTokenCacheSize => {
            Ok(SetupOption::MaxAuthTokenCacheSize { max_size: *value })
          }
          _ => Err(ParseError::KeyValueFormattingError {
            context: "SetupOption::deserialize",
          }),
        }
      }
      KeyValuePair::Bytes { type_value, value } => {
        let type_value = SetupOptionType::try_from(*type_value)?;
        let mut payload_bytes = value.clone();
        match type_value {
          SetupOptionType::Path => {
            let moqt_path =
              String::from_utf8(value.to_vec()).map_err(|e| ParseError::InvalidUTF8 {
                context: "SetupOption::deserialize",
                details: e.to_string(),
              })?;
            Ok(SetupOption::Path { moqt_path })
          }
          SetupOptionType::AuthorizationToken => Ok(SetupOption::new_auth_token(
            AuthorizationToken::deserialize(&mut payload_bytes)?,
          )),
          SetupOptionType::MoqtImplementation => {
            let info = String::from_utf8(value.to_vec()).map_err(|e| ParseError::InvalidUTF8 {
              context: "SetupOption::deserialize",
              details: e.to_string(),
            })?;
            Ok(SetupOption::MoqtImplementation { info })
          }
          SetupOptionType::Authority => {
            let authority =
              String::from_utf8(value.to_vec()).map_err(|e| ParseError::InvalidUTF8 {
                context: "SetupOption::deserialize",
                details: e.to_string(),
              })?;
            Ok(SetupOption::Authority { authority })
          }
          _ => Err(ParseError::KeyValueFormattingError {
            context: "SetupOption::deserialize",
          }),
        }
      }
    }
  }
}

impl TryInto<KeyValuePair> for SetupOption {
  type Error = ParseError;
  fn try_into(self) -> Result<KeyValuePair, Self::Error> {
    match self {
      SetupOption::Path { moqt_path } => {
        let data = moqt_path.as_bytes();
        KeyValuePair::try_new_bytes(SetupOptionType::Path as u64, Bytes::copy_from_slice(data))
      }
      SetupOption::AuthorizationToken { token } => KeyValuePair::try_new_bytes(
        SetupOptionType::AuthorizationToken as u64,
        token.serialize()?,
      ),
      SetupOption::MaxAuthTokenCacheSize { max_size } => {
        KeyValuePair::try_new_varint(SetupOptionType::MaxAuthTokenCacheSize as u64, max_size)
      }
      SetupOption::MoqtImplementation { info } => KeyValuePair::try_new_bytes(
        SetupOptionType::MoqtImplementation as u64,
        Bytes::copy_from_slice(info.as_bytes()),
      ),
      SetupOption::Authority { authority } => KeyValuePair::try_new_bytes(
        SetupOptionType::Authority as u64,
        Bytes::copy_from_slice(authority.as_bytes()),
      ),
    }
  }
}

#[cfg(test)]
mod tests {
  use super::SetupOption;
  use crate::model::common::pair::KeyValuePair;
  use crate::model::common::varint::BufMutVarIntExt;
  use crate::model::parameter::constant::SetupOptionType;
  use bytes::{Buf, BytesMut};

  #[test]
  fn test_roundtrip_path() {
    let orig = SetupOption::new_path("test/path".to_string());
    let serialized = orig.serialize().unwrap();
    let mut buf = serialized.clone();
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(orig, got);
    assert_eq!(buf.remaining(), 0);
  }

  #[test]
  fn test_roundtrip_empty_path() {
    let orig = SetupOption::new_path(String::new());
    let serialized = orig.serialize().unwrap();
    let mut buf = serialized.clone();
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(orig, got);
    assert_eq!(buf.remaining(), 0);
  }

  #[test]
  fn test_roundtrip_max_auth_cache_size() {
    let orig = SetupOption::new_max_auth_token_cache_size(123456);
    let serialized = orig.serialize().unwrap();
    let mut buf = serialized.clone();
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(orig, got);
    assert_eq!(buf.remaining(), 0);
  }

  #[test]
  fn test_roundtrip_moqt_implementation() {
    let orig = SetupOption::new_moqt_implementation("moqtail".to_string());
    let serialized = orig.serialize().unwrap();
    let mut buf = serialized.clone();
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(orig, got);
    assert_eq!(buf.remaining(), 0);
  }

  #[test]
  fn test_roundtrip_authority() {
    let orig = SetupOption::new_authority("example.com:9443".to_string());
    let serialized = orig.serialize().unwrap();
    let mut buf = serialized.clone();
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(orig, got);
    assert_eq!(buf.remaining(), 0);
  }

  #[test]
  fn test_deserialize_invalid_type() {
    // Create a KeyValuePair with an invalid type that SetupOption should reject
    let kvp = KeyValuePair::VarInt {
      type_value: 999, // invalid SetupOptionType
      value: 42,
    };
    let err = SetupOption::deserialize(&kvp);
    assert!(err.is_err());
  }

  #[test]
  fn test_deserialize_path_missing_length() {
    let mut buf = BytesMut::new();
    buf.put_vi(SetupOptionType::Path as u64).unwrap();
    // no length, no data
    let mut bytes = buf.freeze();
    let err = KeyValuePair::deserialize(&mut bytes);
    assert!(err.is_err());
  }

  #[test]
  fn test_deserialize_path_insufficient_data() {
    let mut buf = BytesMut::new();
    buf.put_vi(SetupOptionType::Path as u64).unwrap();
    buf.put_vi(5).unwrap(); // declare 5 bytes
    buf.extend_from_slice(b"abc"); // only 3 bytes
    let mut bytes = buf.freeze();
    let err = KeyValuePair::deserialize(&mut bytes);
    assert!(err.is_err());
  }

  #[test]
  fn test_deserialize_max_auth_cache_missing_value() {
    let mut buf = BytesMut::new();
    buf
      .put_vi(SetupOptionType::MaxAuthTokenCacheSize as u64)
      .unwrap();
    // no size varint
    let mut bytes = buf.freeze();
    let err = KeyValuePair::deserialize(&mut bytes);
    assert!(err.is_err());
  }

  #[test]
  fn test_excess_bytes_after_path() {
    let orig = SetupOption::new_path("ok".to_string());
    let mut buf = BytesMut::from(&orig.serialize().unwrap()[..]);
    buf.extend_from_slice(b"XYZ");
    let mut bytes = buf.freeze();
    let kvp = KeyValuePair::deserialize(&mut bytes).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(got, orig);
    assert_eq!(bytes.remaining(), 3);
    assert_eq!(&bytes[..], b"XYZ");
  }

  #[test]
  fn test_excess_bytes_after_max_auth_cache() {
    let orig = SetupOption::new_max_auth_token_cache_size(7);
    let mut buf = BytesMut::from(&orig.serialize().unwrap()[..]);
    buf.extend_from_slice(&[1, 2, 3]);
    let mut bytes = buf.freeze();
    let kvp = KeyValuePair::deserialize(&mut bytes).unwrap();
    let got = SetupOption::deserialize(&kvp).unwrap();
    assert_eq!(got, orig);
    assert_eq!(bytes.remaining(), 3);
    assert_eq!(&bytes[..], &[1, 2, 3]);
  }
}

File: libs/moqtail-ts/src/model/parameter/setup_option.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValuePair } from '../common/pair'
import { Path, MaxAuthTokenCacheSize, Authority, MoqtImplementation } from './setup'
import { AuthorizationToken } from './common'
import { SetupOptionType, TokenAliasType } from './constant'
import { ProtocolViolationError } from '../error/error'

export type SetupOption = Path | MaxAuthTokenCacheSize | AuthorizationToken | Authority | MoqtImplementation
export namespace SetupOption {
  export function fromKeyValuePair(pair: KeyValuePair): SetupOption | undefined {
    return (
      Path.fromKeyValuePair(pair) ||
      MaxAuthTokenCacheSize.fromKeyValuePair(pair) ||
      AuthorizationToken.fromKeyValuePair(pair) ||
      Authority.fromKeyValuePair(pair) ||
      MoqtImplementation.fromKeyValuePair(pair)
    )
  }
  export function toKeyValuePair(param: SetupOption): KeyValuePair {
    return param.toKeyValuePair()
  }
  export function isPath(param: SetupOption): param is Path {
    return param instanceof Path
  }
  export function isMaxAuthTokenCacheSize(param: SetupOption): param is MaxAuthTokenCacheSize {
    return param instanceof MaxAuthTokenCacheSize
  }
  export function isAuthorizationToken(param: SetupOption): param is AuthorizationToken {
    return param instanceof AuthorizationToken
  }
  export function isAuthority(param: SetupOption): param is Authority {
    return param instanceof Authority
  }
  export function isMoqtImplementation(param: SetupOption): param is MoqtImplementation {
    return param instanceof MoqtImplementation
  }
}

/** Throws if `params` carries AUTHORITY, which MUST NOT be sent over WebTransport. */
export function assertNoAuthorityOverWebTransport(params: KeyValuePair[]): void {
  if (params.some((p) => Number(p.typeValue) === SetupOptionType.Authority)) {
    throw new ProtocolViolationError(
      'assertNoAuthorityOverWebTransport',
      'AUTHORITY setup option MUST NOT be sent over WebTransport (draft-18 §10.3.1.1)',
    )
  }
}

export class SetupOptions {
  private kvps: KeyValuePair[] = []

  addMaxAuthTokenCacheSize(maxSize: bigint | number): this {
    this.kvps.push(new MaxAuthTokenCacheSize(BigInt(maxSize)).toKeyValuePair())
    return this
  }

  addPath(moqtPath: string): this {
    this.kvps.push(new Path(moqtPath).toKeyValuePair())
    return this
  }

  addAuthorizationToken(auth: AuthorizationToken): this {
    this.kvps.push(auth.toKeyValuePair())
    return this
  }

  addMoqtImplementation(info: string): this {
    this.kvps.push(new MoqtImplementation(info).toKeyValuePair())
    return this
  }

  addRaw(pair: KeyValuePair): this {
    this.kvps.push(pair)
    return this
  }

  build(): KeyValuePair[] {
    return this.kvps
  }

  static fromKeyValuePairs(kvps: KeyValuePair[]): SetupOption[] {
    const result: SetupOption[] = []
    for (const kvp of kvps) {
      const parsed =
        Path.fromKeyValuePair(kvp) ||
        MaxAuthTokenCacheSize.fromKeyValuePair(kvp) ||
        AuthorizationToken.fromKeyValuePair(kvp) ||
        Authority.fromKeyValuePair(kvp) ||
        MoqtImplementation.fromKeyValuePair(kvp)
      if (parsed) result.push(parsed)
    }
    return result
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('SetupOptions', () => {
    test('build and fromKeyValuePairs returns correct parameters', () => {
      const kvps = new SetupOptions()
        .addPath('abc')
        .addMaxAuthTokenCacheSize(123n)
        .addAuthorizationToken(AuthorizationToken.newUseAlias(1n))
        .addRaw(new Authority('example.com').toKeyValuePair())
        .addMoqtImplementation('moqtail-ts/0.1')
        .build()
      const parsed = SetupOptions.fromKeyValuePairs(kvps)
      expect(parsed.length).toBe(5)
      expect(parsed[0] && SetupOption.isPath(parsed[0]) && parsed[0].moqtPath === 'abc').toBe(true)
      expect(parsed[1] && SetupOption.isMaxAuthTokenCacheSize(parsed[1]) && parsed[1].maxSize === 123n).toBe(true)
      expect(
        parsed[2] &&
          SetupOption.isAuthorizationToken(parsed[2]) &&
          parsed[2].variant.aliasType === TokenAliasType.UseAlias,
      ).toBe(true)
      expect(parsed[3] && SetupOption.isAuthority(parsed[3]) && parsed[3].authority === 'example.com').toBe(true)
      expect(parsed[4] && SetupOption.isMoqtImplementation(parsed[4]) && parsed[4].info === 'moqtail-ts/0.1').toBe(true)
    })
    test('fromKeyValuePairs skips unknown parameter', () => {
      const unknown = KeyValuePair.tryNewVarInt(998, 1n)
      const kvps = new SetupOptions().addRaw(unknown).addPath('wololoo').build()
      const parsed = SetupOptions.fromKeyValuePairs(kvps)
      expect(parsed.length).toBe(1)
      expect(parsed[0] && SetupOption.isPath(parsed[0]) && parsed[0].moqtPath === 'wololoo').toBe(true)
    })
  })

  describe('assertNoAuthorityOverWebTransport', () => {
    test('throws when params carry AUTHORITY', () => {
      const params = new SetupOptions().addPath('/x').addRaw(new Authority('example.com').toKeyValuePair()).build()
      expect(() => assertNoAuthorityOverWebTransport(params)).toThrow(ProtocolViolationError)
    })
    test('does not throw when params carry no AUTHORITY', () => {
      const params = new SetupOptions().addPath('/x').addMaxAuthTokenCacheSize(1n).build()
      expect(() => assertNoAuthorityOverWebTransport(params)).not.toThrow()
    })
    test('does not throw for an empty params list', () => {
      expect(() => assertNoAuthorityOverWebTransport([])).not.toThrow()
    })
  })
}
